include/vast/Util/Pipeline.hpp

```diff
@@ -0,0 +1,187 @@
+// Copyright (c) 2023-present, Trail of Bits, Inc.
+
+#pragma once
+
+#include "vast/Util/Warnings.hpp"
+
+VAST_RELAX_WARNINGS
+#include <llvm/ADT/DenseSet.h>
+#include <mlir/IR/DialectRegistry.h>
+#include <mlir/Pass/PassManager.h>
+#include <mlir/Pass/PassRegistry.h>
+#include <mlir/Pass/Pass.h>
+VAST_UNRELAX_WARNINGS
+
+#include "vast/Util/Common.hpp"
+
+#include "gap/core/generator.hpp"
+
+namespace vast {
+
+    //
+    // pipeline_step is a single step in the pipeline that is a pass or a list
+    // of pipelines
+    //
+    // Each step defiens a list of dependencies, which are scheduled before the step
+    //
+    struct pipeline_step;
+
+    using pipeline_step_ptr = std::unique_ptr< pipeline_step >;
+
+
+    //
+    // pipeline is a pass manager, which keeps track of duplicit passes and does
+    // not schedule them twice
+    //
+    struct pipeline_t : mlir::PassManager
+    {
+        using base      = mlir::PassManager;
+        using pass_id_t = mlir::TypeID;
+
+        using base::base;
+
+        void addPass(std::unique_ptr< mlir::Pass > pass); 
+
+        template< typename parent_t >
+        void addNestedPass(std::unique_ptr< mlir::Pass > pass) {
+            auto id = pass->getTypeID();
+            if (seen.count(id)) {
+                return;
+            }
+
+            seen.insert(id);
+            base::addNestedPass< parent_t >(std::move(pass));
+        }
+
+        friend pipeline_t &operator<<(pipeline_t &ppl, pipeline_step_ptr pass);
+
+        llvm::DenseSet< pass_id_t > seen;
+    };
+
+
+    using pipeline_step_builder = std::function< pipeline_step_ptr(void) >;
+
+    //
+    // initilizer wrapper to setup dependencies after make is called
+    //
+    template< typename step_t >
+    struct pipeline_step_init : pipeline_step_ptr {
+        using pipeline_step_ptr::pipeline_step_ptr;
```

Comment:
Is this even doing anything when we have a `args_t && ...` ctor later? (I think it will catch all invocations).